File: projects/native-apps/ios/App/SloganHandoff.swift
import CoreGraphics
import Foundation

struct TextPose: Equatable {
	let opacity: Double
	let scale: Double
	let blur: Double

	static let shown = condensed(0)
	static let hidden = condensed(1)

	static func condensed(_ amount: Double) -> TextPose {
		TextPose(opacity: 1 - amount, scale: 1 - 0.55 * amount, blur: 6 * amount)
	}
}

struct TextPoses: Equatable {
	let outgoing: TextPose
	let incoming: TextPose
}

struct StarHead: Equatable {
	let center: CGPoint
	let coreRadius: Double
	let glowRadius: Double
	let opacity: Double
	let hue: CosmicHue
}

struct SloganComet: Equatable {
	let strokes: [FilamentStroke]
	let head: StarHead?

	static let idle = SloganComet(strokes: [], head: nil)
}

struct CometPath: Equatable {
	let start: CGPoint
	let control: CGPoint
	let end: CGPoint

	init(from start: CGPoint, to end: CGPoint, bow: Double) {
		self.start = start
		self.end = end
		let chordX = end.x - start.x
		let chordY = end.y - start.y
		control = CGPoint(
			x: (start.x + end.x) / 2 - chordY * bow,
			y: (start.y + end.y) / 2 + chordX * bow
		)
	}

	func point(at u: Double) -> CGPoint {
		let v = 1 - u
		return CGPoint(
			x: v * v * start.x + 2 * v * u * control.x + u * u * end.x,
			y: v * v * start.y + 2 * v * u * control.y + u * u * end.y
		)
	}

	func samples(from lower: Double, to upper: Double, count: Int) -> [CGPoint] {
		(0..<count).map { point(at: lower + (upper - lower) * Double($0) / Double(count - 1)) }
	}
}

struct SloganHandoff: Equatable {
	let outgoing: String
	let incoming: String
	let startedAt: TimeInterval
	let ordinal: Int
	let seed: UInt64

	static let collapseSeconds = 0.30
	static let ascentStart = 0.20
	static let flightSeconds = 0.70
	static let drainSeconds = 0.20
	static let landingAt = ascentStart + flightSeconds
	static let descentStart = landingAt + 0.45
	static let arrivalAt = descentStart + flightSeconds
	static let bloomSeconds = 0.45
	static let duration = arrivalAt + bloomSeconds
	private static let tailSpan = 0.35
	private static let bow = 0.28
	private static let coreRadius = 4.5
	private static let glowRadius = 13.0
	private static let bloomRadius = 64.0
	private static let cometOpacity = 0.85
	private static let cometLane = -1

	var hue: CosmicHue {
		CosmicHue.allCases[Int(unit(.hue) * Double(CosmicHue.allCases.count))]
	}

	func landing(sky: CGRect) -> CGPoint {
		CGPoint(
			x: sky.minX + (0.18 + 0.64 * unit(.landingX)) * sky.width,
			y: sky.minY + (0.15 + 0.60 * unit(.landingY)) * sky.height
		)
	}

	func visit(sky: CGRect) -> StarVisit {
		StarVisit(anchor: landing(sky: sky), bornAt: startedAt + Self.landingAt, hue: hue, ordinal: ordinal)
	}

	func poses(at elapsed: TimeInterval) -> TextPoses {
		let t = elapsed - startedAt
		return TextPoses(
			outgoing: .condensed(smoothstep(clamped(t / Self.collapseSeconds))),
			incoming: .condensed(1 - smoothstep(clamped((t - Self.arrivalAt) / Self.bloomSeconds)))
		)
	}

	func comet(at elapsed: TimeInterval, subtitleCenter: CGPoint, sky: CGRect, launch: CGPoint) -> SloganComet {
		let t = elapsed - startedAt
		if t < 0 || t >= Self.duration {
			return .idle
		}
		if t < Self.ascentStart {
			let spark = smoothstep(t / Self.ascentStart)
			return SloganComet(
				strokes: [],
				head: StarHead(
					center: subtitleCenter,
					coreRadius: Self.coreRadius * spark,
					glowRadius: Self.glowRadius * spark,
					opacity: Self.cometOpacity * spark,
					hue: hue
				)
			)
		}
		let bow = unit(.bow) < 0.5 ? -Self.bow : Self.bow
		if t < Self.descentStart {
			let ascent = CometPath(from: subtitleCenter, to: landing(sky: sky), bow: bow)
			return flight(along: ascent, since: t - Self.ascentStart, headFades: true)
		}
		let descent = CometPath(from: launch, to: subtitleCenter, bow: bow)
		let inFlight = flight(along: descent, since: t - Self.descentStart, headFades: false)
		if t < Self.arrivalAt {
			return inFlight
		}
		let bloom = smoothstep((t - Self.arrivalAt) / Self.bloomSeconds)
		return SloganComet(
			strokes: inFlight.strokes,
			head: StarHead(
				center: subtitleCenter,
				coreRadius: Self.coreRadius,
				glowRadius: Self.glowRadius + (Self.bloomRadius - Self.glowRadius) * bloom,
				opacity: Self.cometOpacity * (1 - bloom),
				hue: hue
			)
		)
	}

	private func flight(along path: CometPath, since: Double, headFades: Bool) -> SloganComet {
		let u = smoothstep(min(1, since / Self.flightSeconds))
		let drain = clamped((since - Self.flightSeconds) / Self.drainSeconds)
		let tailStart = max(0, u - Self.tailSpan * (1 - drain))
		let strokes = u > tailStart
			? FilamentStroke.tapered(
				along: path.samples(from: tailStart, to: u, count: FilamentStroke.sampleCount),
				hue: hue,
				lane: Self.cometLane,
				tone: .star,
				coreOpacity: Self.cometOpacity
			)
			: []
		let headOpacity = Self.cometOpacity * (headFades ? 1 - drain : 1)
		let head = headOpacity > 0
			? StarHead(
				center: path.point(at: u),
				coreRadius: Self.coreRadius,
				glowRadius: Self.glowRadius,
				opacity: headOpacity,
				hue: hue
			)
			: nil
		return SloganComet(strokes: strokes, head: head)
	}

	private func clamped(_ x: Double) -> Double {
		min(1, max(0, x))
	}

	private enum Slot: UInt64 {
		case landingX = 1
		case landingY
		case bow
		case hue
	}

	private func unit(_ slot: Slot) -> Double {
		var z = seed
		z ^= UInt64(ordinal) &* 0x9E37_79B9_7F4A_7C15
		z ^= slot.rawValue &* 0xBF58_476D_1CE4_E5B9
		return splitMix64Unit(z)
	}
}

File: projects/native-apps/android/app/src/main/kotlin/com/readplace/android/app/SloganHandoff.kt
package com.readplace.android.app

/**
 * A subtitle line's visual state: fully shown ([shown]) collapses toward gone
 * ([hidden]) as the words condense into the spark that carries them away.
 */
data class TextPose(
	val opacity: Double,
	val scale: Double,
	val blur: Double,
) {
	companion object {
		/** The words at [amount] of the way condensed: fading, shrinking, and
		 * blurring together in step. */
		fun condensed(amount: Double): TextPose =
			TextPose(opacity = 1 - amount, scale = 1 - 0.55 * amount, blur = 6 * amount)

		val shown: TextPose = condensed(0.0)
		val hidden: TextPose = condensed(1.0)
	}
}

data class TextPoses(
	val outgoing: TextPose,
	val incoming: TextPose,
)

/** The comet's bright point: a solid core inside a softer glow, in the hue of the
 * slogan it is carrying. */
data class StarHead(
	val center: WavePoint,
	val coreRadius: Double,
	val glowRadius: Double,
	val opacity: Double,
	val hue: CosmicHue,
)

/** One frame of the comet: its tapered trail and, while it burns, its head. */
data class SloganComet(
	val strokes: List<FilamentStroke>,
	val head: StarHead?,
) {
	companion object {
		val idle = SloganComet(strokes = emptyList(), head = null)
	}
}

/**
 * A quadratic curve bowed off the straight line between two points, so the comet
 * arcs across the wordmark instead of shooting through it like a bullet.
 */
class CometPath(
	val start: WavePoint,
	val control: WavePoint,
	val end: WavePoint,
) {
	fun point(u: Double): WavePoint {
		val v = 1 - u
		return WavePoint(
			x = v * v * start.x + 2 * v * u * control.x + u * u * end.x,
			y = v * v * start.y + 2 * v * u * control.y + u * u * end.y,
		)
	}

	fun samples(from: Double, to: Double, count: Int): List<WavePoint> =
		(0 until count).map { point(from + (to - from) * it.toDouble() / (count - 1).toDouble()) }

	companion object {
		fun bowed(from: WavePoint, to: WavePoint, bow: Double): CometPath {
			val chordX = to.x - from.x
			val chordY = to.y - from.y
			return CometPath(
				start = from,
				control = WavePoint(
					x = (from.x + to.x) / 2 - chordY * bow,
					y = (from.y + to.y) / 2 + chordX * bow,
				),
				end = to,
			)
		}
	}
}

/**
 * The whole slogan hand-over as pure geometry over one storm-clock timeline: the
 * outgoing words condense into a spark, the spark flies up and strikes the sky,
 * a star leaves that strike, descends, and blooms into the incoming words. Seeded
 * off [seed] and [ordinal] so every hand-over lands somewhere new in its own hue,
 * and deterministic in [elapsed] so the same moment always draws the same frame.
 */
data class SloganHandoff(
	val outgoing: String,
	val incoming: String,
	val startedAt: Double,
	val ordinal: Int,
	val seed: ULong,
) {
	val hue: CosmicHue
		get() = CosmicHue.entries[(unit(Slot.HUE) * CosmicHue.entries.size.toDouble()).toInt()]

	fun landing(sky: WaveRect): WavePoint =
		WavePoint(
			x = sky.x + (0.18 + 0.64 * unit(Slot.LANDING_X)) * sky.width,
			y = sky.y + (0.15 + 0.60 * unit(Slot.LANDING_Y)) * sky.height,
		)

	fun visit(sky: WaveRect): StarVisit =
		StarVisit(anchor = landing(sky), bornAt = startedAt + LANDING_AT, hue = hue, ordinal = ordinal)

	fun poses(elapsed: Double): TextPoses {
		val t = elapsed - startedAt
		return TextPoses(
			outgoing = TextPose.condensed(smoothstep(clamped(t / COLLAPSE_SECONDS))),
			incoming = TextPose.condensed(1 - smoothstep(clamped((t - ARRIVAL_AT) / BLOOM_SECONDS))),
		)
	}

	fun comet(elapsed: Double, subtitleCenter: WavePoint, sky: WaveRect, launch: WavePoint): SloganComet {
		val t = elapsed - startedAt
		if (t < 0 || t >= DURATION) return SloganComet.idle
		if (t < ASCENT_START) {
			val spark = smoothstep(t / ASCENT_START)
			return SloganComet(
				strokes = emptyList(),
				head = StarHead(
					center = subtitleCenter,
					coreRadius = CORE_RADIUS * spark,
					glowRadius = GLOW_RADIUS * spark,
					opacity = COMET_OPACITY * spark,
					hue = hue,
				),
			)
		}
		val bow = if (unit(Slot.BOW) < 0.5) -BOW else BOW
		if (t < DESCENT_START) {
			val ascent = CometPath.bowed(from = subtitleCenter, to = landing(sky), bow = bow)
			return flight(path = ascent, since = t - ASCENT_START, headFades = true)
		}
		val descent = CometPath.bowed(from = launch, to = subtitleCenter, bow = bow)
		val inFlight = flight(path = descent, since = t - DESCENT_START, headFades = false)
		if (t < ARRIVAL_AT) return inFlight
		val bloom = smoothstep((t - ARRIVAL_AT) / BLOOM_SECONDS)
		return SloganComet(
			strokes = inFlight.strokes,
			head = StarHead(
				center = subtitleCenter,
				coreRadius = CORE_RADIUS,
				glowRadius = GLOW_RADIUS + (BLOOM_RADIUS - GLOW_RADIUS) * bloom,
				opacity = COMET_OPACITY * (1 - bloom),
				hue = hue,
			),
		)
	}

	private fun flight(path: CometPath, since: Double, headFades: Boolean): SloganComet {
		val u = smoothstep(minOf(1.0, since / FLIGHT_SECONDS))
		val drain = clamped((since - FLIGHT_SECONDS) / DRAIN_SECONDS)
		val tailStart = maxOf(0.0, u - TAIL_SPAN * (1 - drain))
		val strokes = if (u > tailStart) {
			FilamentStroke.tapered(
				along = path.samples(from = tailStart, to = u, count = FilamentStroke.SAMPLE_COUNT),
				hue = hue,
				lane = COMET_LANE,
				tone = StrokeTone.STAR,
				coreOpacity = COMET_OPACITY,
			)
		} else {
			emptyList()
		}
		val headOpacity = COMET_OPACITY * (if (headFades) 1 - drain else 1.0)
		val head = if (headOpacity > 0) {
			StarHead(
				center = path.point(u),
				coreRadius = CORE_RADIUS,
				glowRadius = GLOW_RADIUS,
				opacity = headOpacity,
				hue = hue,
			)
		} else {
			null
		}
		return SloganComet(strokes = strokes, head = head)
	}

	private fun clamped(x: Double): Double = minOf(1.0, maxOf(0.0, x))

	private enum class Slot(val raw: ULong) {
		LANDING_X(1uL),
		LANDING_Y(2uL),
		BOW(3uL),
		HUE(4uL),
	}

	private fun unit(slot: Slot): Double {
		var z = seed
		z = z xor (ordinal.toULong() * 0x9E37_79B9_7F4A_7C15uL)
		z = z xor (slot.raw * 0xBF58_476D_1CE4_E5B9uL)
		return splitMix64Unit(z)
	}

	companion object {
		const val COLLAPSE_SECONDS = 0.30
		const val ASCENT_START = 0.20
		const val FLIGHT_SECONDS = 0.70
		const val DRAIN_SECONDS = 0.20
		const val LANDING_AT = ASCENT_START + FLIGHT_SECONDS
		const val DESCENT_START = LANDING_AT + 0.45
		const val ARRIVAL_AT = DESCENT_START + FLIGHT_SECONDS
		const val BLOOM_SECONDS = 0.45
		const val DURATION = ARRIVAL_AT + BLOOM_SECONDS

		private const val TAIL_SPAN = 0.35
		private const val BOW = 0.28
		private const val CORE_RADIUS = 4.5
		private const val GLOW_RADIUS = 13.0
		private const val BLOOM_RADIUS = 64.0
		private const val COMET_OPACITY = 0.85
		private const val COMET_LANE = -1
	}
}
